### PR TITLE
Add helper functions for namespaced topic resolution and unit test setup

### PR DESCRIPTION
**Pull request by C88-YQ**
# 🎉 New feature

Related to #3878

## Summary

This PR mainly adds helper functions for topic namespace support and for related tests.

- `resolvedTopicName`
  - Resolves topic names from SDF plugin elements.
  - Uses a non-empty custom topic from SDF when available.
  - Prefixes relative custom topics with the provided namespace.
  - Leaves absolute custom topics unchanged.
  - Falls back to the default topic when the SDF element is missing or empty.
  - Adds tests coverage for relative, absolute, empty, and missing topic elements.

- `UnitTestUtil`
  - Adds reusable helpers for setting up unit test contexts.
  - Supports loading a world from an SDF string and creating its entity.
  - Supports loading a model from an SDF string and creating its entity.
  - Optionally extracts the first plugin SDF for test use.

## Backport Policy
- [ ] This is safe to backport to the following versions:
    - [x] Jetty
    - [x] Ionic
    - [x] Harmonic
    - [ ] Fortress
- [ ] This **should not** be backported
- [ ] I am not sure
- [ ] Other (fill in yourself)

## Test it
```
cd build/gz-sim
ctest
```

## Checklist
- [x] Signed all commits for DCO
- [ ] Added a screen capture or video to the PR description that demonstrates the feature
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [x] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [x] Updated Bazel files (if adding new files). Created an issue otherwise.
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers
- [ ] Was GenAI used to generate this PR? If so, make sure to add "Assisted-by" to your commits. (See [this policy](https://osralliance.org/wp-content/uploads/2025/05/OSRF-Policy-on-the-Use-of-Generative-Tools-Generative-AI-in-Contributions.pdf) for more info.)

### CODE REVIEW

**Review by luca-della-vedova — COMMENTED** [review]
The Bazel CI failure looks real:

```
test/helpers/UnitTestUtil.hh:24:10: error: module //test:Helpers does not depend on a module exporting 'sdf/Root.hh'
   24 | #include <sdf/Root.hh>
      |          ^
test/helpers/UnitTestUtil.hh:25:10: error: module //test:Helpers does not depend on a module exporting 'sdf/World.hh'
   25 | #include <sdf/World.hh>
      |          ^
test/helpers/UnitTestUtil.hh:26:10: error: module //test:Helpers does not depend on a module exporting 'sdf/Model.hh'
   26 | #include <sdf/Model.hh>
      |          ^
test/helpers/UnitTestUtil.hh:27:10: error: module //test:Helpers does not depend on a module exporting 'sdf/Plugin.hh'
   27 | #include <sdf/Plugin.hh>
      |          ^
```

I think you might need to add a dependency to sdformat for the bazel test targets

**Comment by C88-YQ** [comment]
@luca-della-vedova 

Sorry for the late reply.

The original Bazel layering error is fixed by adding `@sdformat` to `//test:Helpers`.

The current CI failure is different and comes from `gz-transport` / `zenoh-cpp`, where `std::optional<zenoh::CancellationToken>` fails to compile with clang/libstdc++. So it appears unrelated to this PR.

File: include/gz/sim/Util.hh [hunk 1]
```diff
@@ -78,6 +78,23 @@ namespace gz
...
+    /// topic is available, the default topic is used.
+    /// \param[in] _sdf SDF to read the topic name from.
+    /// \param[in] _sdfElement Name of the SDF child element containing the
+    /// topic name.
+    /// \param[in] _ns Namespace to prepend to relative topic name.
+    /// \param[in] _defaultTopic Topic to use when no custom topic is specified.
+    /// \return A valid Gazebo Transport topic name, or an empty string if no
+    /// valid topic could be generated.
+    std::string GZ_SIM_VISIBLE resolvedTopicName(
+        const std::shared_ptr<const sdf::Element> &_sdf,
```
Comment (azeey) [inline include/gz/sim/Util.hh:94]: Can you forward declare `sdf::Element` since `sdf/Element.hh` is not directly included here.

File: src/Util.cc [hunk 2]
```diff
@@ -282,6 +282,32 @@ std::string scopedNamespace(const EntityComponentManager &_ecm,
   return result;
 }
 
+std::string resolvedTopicName(const std::shared_ptr<const sdf::Element> &_sdf,
```
Comment (azeey) [inline src/Util.cc:285]: Include `sdf/Element.hh`.

File: include/gz/sim/Util.hh [hunk 3]
```diff
@@ -78,6 +78,23 @@ namespace gz
...
+        const std::string &_sdfElement, const std::string &_ns,
```
Comment (azeey) [inline include/gz/sim/Util.hh:95]:
```suggestion
        const std::string &_sdfElementName, const std::string &_namespace,
```

File: src/Util.cc [hunk 4]
```diff
@@ -282,6 +282,32 @@ std::string scopedNamespace(const EntityComponentManager &_ecm,
...
+  {
+    std::string customTopic = _sdf->Get<std::string>(_sdfElement);
+
+    if (!customTopic.empty())
+    {
+      // Only prepend namespace to relative topic names.
+      // Absolute topic names (starting with '/') are left unchanged.
+      if (customTopic.front() != '/')
+      {
+        customTopic = _ns + "/" + customTopic;
```
Comment (azeey) [inline src/Util.cc:301]: Is it a problem if `_ns` is `/`?

File: src/Util_TEST.cc [hunk 5]
```diff
@@ -308,6 +308,40 @@ TEST_F(UtilTest, ScopedNamespace)
...
+  EXPECT_EQ("/ns/test_relative_topic",
+    resolvedTopicName(sdf, "relative_topic", "/ns", "/default_relative_topic"));
+  EXPECT_EQ("/test_absolute_topic",
+    resolvedTopicName(sdf, "absolute_topic", "", "/default_absolute_topic"));
+  EXPECT_EQ("/test_absolute_topic",
+    resolvedTopicName(sdf, "absolute_topic", "/ns", "/default_absolute_topic"));
+  EXPECT_EQ("/default_empty_topic",
+    resolvedTopicName(sdf, "empty_topic", "", "/default_empty_topic"));
+  EXPECT_EQ("/default_empty_topic",
+    resolvedTopicName(sdf, "empty_topic", "/ns", "/default_empty_topic"));
```
Comment (azeey) [inline src/Util_TEST.cc:338]: Can you add a test for a default empty topic that is relative to ensure that the namespace is used in that case?

File: test/helpers/UnitTestUtil.hh [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+/// \param[in] _eventMgr Event manager used while creating entities.
+/// \param[out] _entity The entity corresponding to the world
+/// \param[out] _pluginSdf Optional copy of the world's first plugin SDF.
+/// If nullptr, no plugin will be stored
+inline void LoadWorldContext(
+  const std::string &_sdfString,
+  EntityComponentManager &_ecm,
+  EventManager &_eventMgr,
+  Entity &_entity,
+  sdf::Plugin *_pluginSdf = nullptr)
```
Comment (azeey) [inline test/helpers/UnitTestUtil.hh:48]: I'm not sure passing a pointer here is correct. If you want the caller's pointer to be updated, then we'll need to pass a double pointer. If you want the caller's plugin object to be updated (sdf::Plugin value copy), then it'd be better to pass by reference. Better yet, just return the value (see https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f20-for-out-output-values-prefer-return-values-to-output-parameters)

File: include/gz/sim/Util.hh [hunk 7]
```diff
@@ -78,6 +78,23 @@ namespace gz
...
+    /// namespace, while absolute topic names are left unchanged. If no custom
+    /// topic is available, the default topic is used.
+    /// \param[in] _sdf SDF to read the topic name from.
+    /// \param[in] _sdfElement Name of the SDF child element containing the
+    /// topic name.
+    /// \param[in] _ns Namespace to prepend to relative topic name.
+    /// \param[in] _defaultTopic Topic to use when no custom topic is specified.
+    /// \return A valid Gazebo Transport topic name, or an empty string if no
+    /// valid topic could be generated.
+    std::string GZ_SIM_VISIBLE resolvedTopicName(
```
Comment (azeey) [inline include/gz/sim/Util.hh:93]: We have three consecutive string parameters here which can make it easy to accidentally swap the arguments at the call site. Consider creating a struct with `string_view` fields.

File: include/gz/sim/Util.hh (hunk 1 above)
Comment (C88-YQ) [reply include/gz/sim/Util.hh:102]: Thanks for the suggestion! Done :smile:

File: src/Util.cc (hunk 2 above)
Comment (C88-YQ) [reply src/Util.cc:286]: Thanks for the suggestion! Done :smile:

File: src/Util.cc (hunk 4 above)
Comment (C88-YQ) [reply src/Util.cc]: Thanks for the reminder! I updated the prefix handling so `/` is preserved without appending another slash.

File: src/Util_TEST.cc (hunk 5 above)
Comment (C88-YQ) [reply src/Util_TEST.cc:338]: @azeey My idea is that `default_topic` should already be fully resolved before calling this function, so `_ns` is not added to `default_topic` here.

The main reason is to keep the old behavior. Before namespace support, `/model/<model_name>` was only added for default topics, not for user-defined topics. If we use `/model/<model_name>` as `_ns` when no namespace is set, it would also affect user-defined relative topics and change the old behavior.

I see two possible approaches:

1. Resolve the full `default_topic` before calling this function, which is what I currently do.
2. Pass two namespace/prefix values: one for the explicit namespace and one for the default topic prefix.

I chose the first approach, so I didn't add a test for a relative `default_topic` here. Do you think the second approach would be better?

File: include/gz/sim/Util.hh (hunk 3 above)
Comment (C88-YQ) [reply include/gz/sim/Util.hh]: Thanks for the suggestion! Done :smile:

File: test/helpers/UnitTestUtil.hh (hunk 6 above)
Comment (C88-YQ) [reply test/helpers/UnitTestUtil.hh]: Thanks for the suggestion! I changed `_pluginSdf` to be passed by reference since the caller's plugin object is updated here. I also added overloads for the cases where the plugin SDF isn't needed.

File: include/gz/sim/Util.hh (hunk 7 above)
Comment (C88-YQ) [reply include/gz/sim/Util.hh:101]: Thanks for the suggestion! That makes sense.

I ended up using `std::string` for the fields because the SDF APIs used here (`HasElement` and `Get`) take `std::string`, so using `std::string_view` would require additional conversions inside the function.

Is there any case here where `std::string_view` would still be preferable that I might be overlooking?

File: src/Util_TEST.cc (hunk 5 above)
Comment (C88-YQ) [reply src/Util_TEST.cc]: I also saw your related comment in #3878 about handling this convention inside `resolvedTopicName` instead of in each system. I think that makes sense.

I updated `resolvedTopicName` to also take a `defaultTopicPrefix` and handle the prefix selection internally. It now uses the namespace when provided, otherwise it falls back to the original default prefix.

I also added test coverage for a relative default topic.

File: include/gz/sim/Util.hh (hunk 7 above)
Comment (azeey) [reply include/gz/sim/Util.hh:115]: I was hoping to avoid copies when creating the struct. But since all the SDF APIs need `std::string`, what you have seems good.
